rtCommon/bidsInterface.py

```diff
@@ -1,43 +1,292 @@
+        return stream.getNumVolumes()
+
+
+class DicomToBidsStream():
+    """
+    A class that watches for DICOM file creation in a specified directory and with
+    a specified file pattern. When DICOM files arrive it converts them to BIDS
+    incrementals and returns the BIDS incremental. This lets a real-time classification
+    script process data directly as BIDS as it arrives from the scanner.
+    """
+    def __init__(self, dicomDir, dicomFilePattern, dicomMinSize, **entities):
+        """
+        Intialize a new DicomToBids stream, watches for Dicoms and streams as BIDS
+        Args:
+            dicomDir: The directory where the scanner will write new DICOM files
+            dicomFilePattern: A regex style pattern of the DICOM filenames to
+                watch for. They should include a {TR} tag with optional formatting.
+                For example filenames like '001_000013_000005.dcm' would have a
+                pattern '001_000013_{TR:06d}.dcm' where the volume number (TR)
+                will be filled in by a 6 digit leading zeros value.
+            dicomMinSize: Minimum size of the file to return (will continue waiting
+                if below this size)
+            entities: BIDS entities (subject, session, task, run, suffix, datatype) that
+                define the particular subject/run of the data to stream
+        """
+        # TODO - make sure dicomPattern has {TR} in it
+        # TODO - restrict allowed directories, check that dicomDir is in allowed dir
+        if 'suffix' not in entities.keys():
+            entities['suffix'] = 'bold'
+        if 'datatype' not in entities.keys():
+            entities['datatype'] = 'func'
+        self.entities = entities
+        self.dicomDir = dicomDir
+        self.dicomFilePattern = dicomFilePattern
+        self.dataInterface = DataInterface(dataRemote=False, allowedDirs=['*'], allowedFileTypes=['.dcm'])
+        self.dicomStreamId = self.dataInterface.initScannerStream(dicomDir, dicomFilePattern, dicomMinSize)
+        self.initialized = True
+
+    def getNumVolumes(self) -> int:
+        """
+        Return the number of brain volumes in the run, unknowable by this
+        interface ahead of time for a real-time DICOM stream
+        """
+        raise NotImplementedError('getNumVolumes not implemented for DicomBidsStream')
+
+    def getIncremental(self, volIdx) -> BidsIncremental:
+        """
+        Get the BIDS incremental for the corresponding DICOM image indicated
+            by the volIdx, where volIdx is equivalent to TR id.
+        Args:
+            volIdx: The volume index (or TR) within the run to retrieve.
+        Returns:
+            BidsIncremental for the matched DICOM for the run/volume
+        """
+        # wait for the dicom and create a bidsIncremental
+        dcmImg = self.dataInterface.getImageData(self.dicomStreamId, volIdx)
+        dicomMetadata = getDicomMetadata(dcmImg)
+        dicomMetadata.update(self.entities)
```

Comment:
It seems like we could get to this point without  having all required entities for BIDS Incremental in the `entities` dictionary -- perhaps it would be good add to some validation earlier to make sure the streams can't be created without having all of the required metadata (e.g., subject and task)?